Fixes Tooltips Showing Outside Minimum Time Bounds

MPierre9 opened the pull request:
Fixes #5228 by checking if x is outside the minimum border of the graph then allowing no tooltips to be created if so. 

[JSFiddle](https://jsfiddle.net/mqL0jrr2/3/)

## Result 
![tooltip](https://user-images.githubusercontent.com/27080760/36459894-d979c89c-1683-11e8-8a41-1cd2fea29723.gif)

MPierre9 commented:
Travis CI test fails even if I comment everything else out and leave variable `border`.

etimberg commented on src/core/core.tooltip.js:
```diff
@@ -505,8 +505,11 @@ module.exports = function(Chart) {
-				for (i = 0, len = active.length; i < len; ++i) {
-					tooltipItems.push(createTooltipItem(active[i]));
+				var border = active[0]._xScale.left + 7; // 7 used for graph offset
+				if (existingModel.x > border) {
```
I think we should check that `existingModel.x > chartArea.left && existingModel.y < chartArea.right`

MPierre9 replied:
I think that's a great idea I just keep getting `chartArea` is not defined when I use it. I think chartArea
 only exists in `function determineAlignment(tooltip, size)`. 

![chartarea](https://user-images.githubusercontent.com/27080760/36516577-abd34552-174c-11e8-930a-c5d6bbf8d762.JPG)

etimberg commented:
Does the test fail if `border` is set to 0 always? It looks like the tooltip is at a different position after these tests

MPierre9 commented:
It seems to pass only if I comment out the `if(existingModel.x > border)` and leave border set to 0. If I keep the `if` I get 20 failed tests with `gulp test`. I'm not sure what the code is interfering with even with border set to zero it shouldn't obstruct the creation of a tooltip

MPierre9 commented:
After some more testing it seems to be `active[0]._xScale.left` not being available when testing for certain charts like donut and bubble charts. Is there a way to check what chart type is being used in `core.tooltip.js`.

MPierre9 commented:
Updated [JSFiddle](https://jsfiddle.net/ygzqtxny/1/)

I changed the location of the if statement and made a global variable so getting the chartArea would be easier.

etimberg commented on src/core/core.tooltip.js:
```diff
@@ -845,7 +847,7 @@ module.exports = function(Chart) {
 			if (e.type === 'mouseout') {
 				me._active = [];
-			} else {
+			} else if (e.x > globalChartArea) {
```
The global variable will cause problems when there are multiple charts on the same page that could potentially have different sizes. Is there a reason that `me._chart.chartArea` doesn't work?

MPierre9 replied:
Actually `me._chart.chartArea` works, I wasn't sure how to call it. Thanks!

etimberg commented on src/core/core.tooltip.js:
```diff
@@ -845,7 +845,7 @@ module.exports = function(Chart) {
 			if (e.type === 'mouseout') {
 				me._active = [];
-			} else {
+			} else if (e.x > me._chart.chartArea.left) {
```
looks good! One last question, should this also check `e.x < me._chart.chartArea.right`? if so, i'd move `me._chart.chartArea` out into a local variable for an improvement to minification

MPierre9 replied:
Good idea! I added it to the latest PR

MPierre9 replied:
Right now with the right side of the chart the bug doesn't occur but it could be a good fail safe just in case

simonbrunel commented on src/core/core.tooltip.js:
```diff
@@ -839,13 +839,14 @@ module.exports = function(Chart) {
 			var me = this;
 			var options = me._options;
 			var changed = false;
+			var border = me._chart.chartArea;
```
I would rename this variable `area`

MPierre9 replied:
Changed in the latest PR

simonbrunel commented on src/core/core.tooltip.js:
```diff
@@ -839,13 +839,14 @@ module.exports = function(Chart) {
 			if (e.type === 'mouseout') {
 				me._active = [];
-			} else {
+			} else if (e.x > border.left && e.x < border.right) {
```
I think limits should be included because values can be positioned at the start and end pixels.

Shouldn't we do the same for `e.y` but also clear active elements if mouse outside?

```javascript
if (e.type === 'mouseout' ||
    e.x < area.left || e.x > area.right ||
    e.y < area.top || e.y > area.bottom) {
    me._active = []
} else {
    me._active = me._chart.getElementsAtEventForMode(...
}
```

MPierre9 replied:
Good idea I forgot about accounting for y. I'll make the changes, thanks!

MPierre9 commented:
[Latest fiddle](https://jsfiddle.net/28syrfwy/)

## Result 
![chartjstooltip](https://user-images.githubusercontent.com/27080760/36947614-b09a28be-1f9c-11e8-896b-2e48fb575027.gif)


@simonbrunel Seems like the latest commit with your suggestion fixed the issue with the tooltips not showing. How should I go about making unit tests for this fix? I haven't made one for this project yet. Thanks for the help!

MPierre9 commented:
When I do gulp test I'm getting ` Core.Tooltip index mode Should only use x distance when intersect is false FAILED` Thoughts?

simonbrunel commented:
@MPierre9 tests are failing because the triggered [mouse event is outside the chart area](/chartjs/Chart.js/blob/master/test/specs/core.tooltip.tests.js#L71). I would really prefer to update the unit tests with correct mouse position and keep `e.y < area.top || e.y > area.bottom`.

MPierre9 commented:
@simonbrunel Ok I'll look into rewriting the unit test

MPierre9 commented:
@simonbrunel I updated the unit test so it can pass with `e.y < area.top || e.y > area.bottom`. I'm not sure if it is correct. A quick review would be much appreciated. Thanks!

MPierre9 commented:
Sure thing! @benmccann

MPierre9 commented:
Unfortunately ran into too many problems with git so I made another PR with the updates. #5419
